### DISCUSSION TITLE
Would it be possible to use project natures?

### DISCUSSION THREAD

**Discussion by ahmedneilhussain** [eclipse/lsp4e#384] [Ideas]
I discovered when coming up with my proposed API that the way we get the language server for a project is a little unsatisfactory. The implementation is really based around file [content] types, and while we can figure out that no appropriate language server is running and launch one when a file for a given type is opened, we don't do that for projects.

I think this might be a weakness of the LSP itself - it doesn't really have a concept of a project type. When we try and find a LS for a project the best we can do is scan the LS that have started so far [restarting them if necessary] to see if any are associated with a project.

That means if you start eclipse with a project open but no open editors, then Find Workspace Symbols `ctrl-shift-T` won't work, but will start working as soon as you open a file in that project of a content type associated with a LS.

Would it be possible/useful to add an optional parameter to our Language Server Definition extension point that took a project nature? Project-level LS operations on that project could then key off the project nature, even if the LS hadn't started.

What do people think? Maybe it's not worth the trouble since there aren't many project-level (rather than document-level) operations, and having no open documents (and none ever opened) is a bit of an artificial edge-case.

**Comment by mickaelistria** [comment]
So basically a `natureMapping` additionally to existing `contentTypeMapping` that would automatically start or bind a language server for every project that has such nature? Why not.
However, langauge servers do receive all `workspaceFolders` so they need to do project type identification on server side already; so I'm not certain repeating that on the client side is so useful. What about just starting the language server(s) you're interested in by default? That should be enough to get `Ctrl+Shift+T` working without requiring a new way of defining a mapping.

**Comment by ghentschke** [comment]
> Would it be possible/useful to add an optional parameter to our Language Server Definition extension point that took a project nature? Project-level LS operations on that project could then key off the project nature, even if the LS hadn't started.

I would rather appreciate such a solution since in our project there is a requirement that it should be possible to define a LS per project. 

> That means if you start eclipse with a project open but no open editors, then Find Workspace Symbols ctrl-shift-T won't work, but will start working as soon as you open a file in that project of a content type associated with a LS.

Same issue here. I would't start all language servers by default for performance reasons.  
Is there any progress on this topic?

**Reply by ahmedneilhussain** [reply]
I have been a bit preoccupied with non-Eclipse stuff recently and haven't got anywhere with this. I don't think it'd be that hard, although I'm not that much of an expert on Eclipse APIs and extension points.
